README.md
# perriera/extras_mon

Welcome to the perriera/extras_mon project.

[![CMake](https://github.com/mattcoding4days/extras/actions/workflows/cmake.yml/badge.svg?branch=dev)](https://github.com/mattcoding4days/extras/actions/workflows/cmake.yml)

<div align="center">
  <img width="442" height="320" src="assets/extras.png">
  <br>
</div>

## Who would want to use this?

> This project aims to get you up and running with C++17 based project with a 'git clone' operation. As well, this particular C++ project template comes with builtin [CPM](https://github.com/cpm-cmake/CPM.cmake) support. This means that you can incorporate other [github](https://github.com) projects, (or anything that uses the git tool) into your project with the simple inclusion of a CPM directive in in your CMakeLists.txt file.

## Prerequisites

To use this template you will need a Linux environment, (Ubuntu is idea).

1.  [Install pip](https://linuxize.com/post/how-to-install-pip-on-ubuntu-18.04/), (either pip 2 or pip 3)
2.  [Install cmake](https://cmake-python-distributions.readthedocs.io/en/latest/installation.html) using pip, (cmake 3.2.1 or better required)
3.  Create project on [Github.com](https://github.com) using **perriera/extras_mon** out of the box, (or as a Github template).

Once you have established these THREE STEPS you may do the following

    git clone https://github.com/<your project>
    cd <your project>
    mkdir build
    cd build
    cmake ..
    make
    cd ..
    build/run-unittests-extras_mon

If you have done everything correctly, you will see a Catch2 diagnostic with all green, (similar to the following):

    perry@ubuntu:~/Projects/extras_mon$ build/run-unittests-extras_mon
    ===============================================================================
    All tests passed (1 assertion in 2 test cases)

**Recommended**: The project now has been successfully created in git already. So add your first github tag:

    git tag -a v0.1.0 -m "Created extras_mon"
    git push origin v0.1.0

However, you not done yet:

Inside your favorite editor, (or search & replace command utility):

1.  Global search and replace **extras_mon** to <project_name> (case sensitive)
2.  Global search and replace **EXTRAS_MON** to <PROJECT_NAME> (case sensitive)
3.  Change directory name '**include/extras_mon**' to '**include/<project_name>**'
4.  Change the links found in changelog.md such that they begin at 0.1.0 for your project.

For example, instead of:

    [Unreleased]: https://github.com/perriera/extras_mon/compare/v0.7.0...HEAD
    [0.7.0]: https://github.com/perriera/extras_mon/compare/v0.6.6...v0.7.0
    [0.6.6]: https://github.com/perriera/extras_mon/compare/v0.5.6...v0.6.0
    [0.5.6]: https://github.com/perriera/extras_mon/compare/v0.5.5...v0.5.6
    [0.5.5]: https://github.com/perriera/extras_mon/compare/v0.5.4...v0.5.5
    [0.5.4]: https://github.com/perriera/extras_mon/compare/v0.5.3...v0.5.4
    [0.5.3]: https://github.com/perriera/extras_mon/compare/v0.5.2...v0.5.3
    [0.5.2]: https://github.com/perriera/extras_mon/compare/v0.5.1...v0.5.2
    [0.5.1]: https://github.com/perriera/extras_mon/compare/v0.5.0...v0.5.1
    [0.5.0]: https://github.com/perriera/extras_mon/compare/v0.4.2...v0.5.0
    [0.4.2]: https://github.com/perriera/extras_mon/compare/v0.4.1...v0.4.2
    [0.4.1]: https://github.com/perriera/extras_mon/compare/v0.4.0...v0.4.1
    [0.4.0]: https://github.com/perriera/extras_mon/compare/v0.3.1...v0.4.0
    [0.3.1]: https://github.com/perriera/extras_mon/compare/v0.3.0...v0.3.1
    [0.3.0]: https://github.com/perriera/extras_mon/compare/v0.2.0...v0.3.0
    [0.2.0]: https://github.com/perriera/extras_mon/compare/v0.1.0...v0.2.0
    [0.1.0]: https://github.com/perriera/extras_mon/releases/tag/v0.1.0

Make it more like:

    [Unreleased]: https://github.com/perriera/extras_mon/compare/v0.2.0...HEAD
    [0.2.0]: https://github.com/perriera/extras_mon/compare/v0.1.0...v0.2.0
    [0.1.0]: https://github.com/perriera/extras_mon/releases/tag/v0.1.0

Also, while you are at, remove all entries to tags in changelog.md from **[0.2.0]** onward. As in there should only be two tag entries in your new project.

Change this:

    #### 0.8.0 (2021-09-03)
    ### Moved
    - Vendor files

    ...

    #### 0.3.1 (2021-09-03)
    ### Removed
    - Renamed identifier

    #### 0.3.0 (2021-09-03)
    ### Removed
    - extras_mon source code

    #### 0.2.0 (2021-09-03)
    ### Added
    - SHARED library support
    - (perriera) extras library support, (3.11.0)
    - CPM.cmake support

To this:

    #### 0.2.0 (2021-09-03)
    ### Added
    - SHARED library support
    - (perriera) extras library support, (3.11.0)
    - CPM.cmake support

Then change the version information in **include/version.h** to 0.2.0

    #define EXTRAS_MON_VER_MAJOR 0
    #define EXTRAS_MON_VER_MINOR 2
    #define EXTRAS_MON_VER_PATCH 0

**Recommended**: You must make sure that the github tags you setup are always the same as the version number for the project. This has to be done by you manually, (at this point in time).

Then do the following:

    rm -rf build
    cd build
    make
    cd ..

This will re-create the project now with all references to extras_mon changed to extras_mon. To test this do the following:

    build/run-unittests-<press tab>

**Recommended**: The project now has been successfully created in git already. So add your first github tag:

    git tag -a v0.2.0 -m "Renamed C++ template to: extras_mon"
    git push origin v0.2.0

## Changelog.md & include/version.hpp

Remember, when you go to make changes to your changelog.md you should maintain the version number of your project using 'include/version.hpp'. For example, in your version.hpp will look something similiar to the following but all referenced to EXTRAS_MON will be changed to extras_mon

**Recommended**: Inside your Github.com for your project, whenever you add a new tag, go the distance and make it a _release_ as well. Just go to your project on Github, select **tag**. Once the tags page is displayed left mouse click on the three dots you see for the newly added tag and select "Create Release"

## CPM Optimization

> Note: Be sure to have the desired **VERSION** of perrier/extras specified in the **VERSION** keyword of the CPMAddPackage specification, (in your CMakeLists.txt). Also, to optimize CPM support on your projects be sure to set the CPM_SOURCE_CACHE environment variable for shared CPM libraries. Typically you would add this to your `.bashrc` script. With this environment variable set, the CPM package manager will not need to reload after successive build/ directory resets, (aka. `rm -rf build`).

    export CPM_SOURCE_CACHE=$HOME/.cache/CPM

## CPM: Debian package, (Shared libraries) & gdb, (debugging) issue

**NOTE:** If you ever run into a situation where your debugging environment is not displaying the same variable names or code changes that you are currently working on be sure to check whether or not the source code you are working is not currently installed as a Debian package, (as a shared library).

    cd build
    sudo dpkg -r <package-name>

**Further**: Whenever your development environment appears to be acting strange, (as in the compiled code does not appear to be consistent with the source code) remember these three factors:

1.  That you can always backtrack your changes using your source repository.
2.  As well as reset your development environment simply by deleting your **build/** directory.
3.  Also, be sure to keep in mind whether the code you are working on is currently installed as a shared library.

## Visual Studio Code (build/ directory) Compilation Fail Safe

The industry practice of compiling to a build/ directory is supported here. When you detect some funny business going on with your development environment, (for example if the cmake utility keeps rebuilding everything), merely delete the build/ directory, (aka. `rm -rf build`) and recreate it. If you are using the Visual Studio Code environment you can press **SHIFT-CTRL-B**, (see `.vscode/tasks.conf`) which will force a rebuild of the directory, (including re-creating `build/` for you).

> Please note: At present whenever you re-open your project with Visual Studio Code the cmake utility may or may not be in sync with the projects, (you'll notice this when all the source files keep getting rebuilt). To correct this open a Terminal window **inside** the Visual Studio Code environment and delete the build/ directory, (aka. `rm -rf build`)then rebuild the project with **SHIFT-CTRL-B**.

## FakeIt2 C++ Mocking Framework

This project is already setup for use with [FakeIt](https://github.com/eranpeer/FakeIt). Anyone familiar with writing interface mocks, (as in [Mock Object](https://en.wikipedia.org/wiki/Mock_object) should find the built-in support for mocking interfaces using the FakeIt framework to be indispensible here).

## Catch2 C++ Test Case Framework

This project is already setup for use with [Catch2](https://github.com/catchorg/Catch2). Anyone familiar with writing test cases, (as in [TDD - Test Driven Development](https://en.wikipedia.org/wiki/Test-driven_development) should find the built-in support for Catch2 to be indispensible here).

## JIRA/Confluence/Bitbucket support

This project is already setup for use with [Jira (software)](<https://en.wikipedia.org/wiki/Jira_(software)>). Anyone familiar with writing JIRA entries as part of morning [SCRUM](<https://en.wikipedia.org/wiki/Scrum_(software_development)>) meetings, (or bi-weekly [SCRUM SPRINT](https://en.wikipedia.org/wiki/Scrum_Sprint) meetings) will find **perriera/extras_mon** to be well suited to the needs of JIRA/SCRUM/SPRINT.

## Conclusion

Your brand new C++ project is now setup and ready for use with **[perriera/extras](https://github.com/perriera/extras)**, (which comes bundled with [spdlog](https://github.com/gabime/spdlog), [cpr](https://github.com/libcpr/cpr), and [nlohmann json](https://github.com/nlohmann/json) support, plus others).


changelog.md
# Changelog

> All notable changes to this project will be documented in this file.</br>
> The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), </br>
> and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).</br>
> To add a tag to the git repository do:
>
> > git tag -a v0.7.0 -m "Using 4.21.0"</br>
> > git push origin v0.7.0

IMPORTANT:

- Manually keep the changelog version number in sync with CMakeLists.txt project version.<br>
- Keep the changelog.md up to date, neat and tidy, (as it is a convenient tool for code maintenance).<br>
- Use [Semantic Versioning](https://semver.org/spec/v2.0.0.html) such that<br>
- Given a version number MAJOR.MINOR.PATCH, increment the:<br>
- MAJOR version when you make incompatible API changes,<br>
- MINOR version when you add functionality in a backwards compatible manner, and<br>
- PATCH version when you make backwards compatible bug fixes. <br>
- But there is no need to change the version number for correcting spelling errors, (etc).<br>
- MAKE SURE to tag and update the tag compare list below, (for quick code comparison purposes).<br>

## [Unreleased]

Detailed test cases</br>
Detailed class descriptions in README.md</br>
Correct URL for github, (for source compare)</br>

## [Next Steps]

Adding mock test cases </br>
Adding unit test cases </br>
Updated README.md accordinging</br>

#### 0.7.0 (2021-12-22)

### Using

- extras_rsi v4.21.0

#### 0.6.0 (2021-12-22)

### Using

- extras_rsi v4.19.0

#### 0.5.0 (2021-12-22)

### Using

- extras_rsi v4.18.0

#### 0.4.0 (2021-12-22)

### Added

- MonitorInterface

#### 0.3.0 (2021-12-21)

### Using

- extras_rsi v4.16.0

#### 0.2.0 (2021-12-21)

### Renamed

- Named project

#### 0.1.0 (2021-08-26)

### Added

- Fetch_Content support
- changelog.md
- Project VERSION inside CMakeLists.txt & changelog.md
- Adopted new changelog format.
- sudo make install parameters
- shared library support
- sudo checkinstall support
- GoF: Abstract design pattern support
- Debian package support
- extra/interfaces.hpp support
- Unit test casses support, (Catch2)
- Mock framework support, (FakeIt)

### Changed

- CMakeLists.txt revamped & streamlined
- README.md revamped & streamlined
- Added documentation on code
- cxx_std_17 on all targets
- std::filesystem on all targets

### Maintained

- @brief style formatting
- Doxygen support

### Removed

- Conan package manager deprecated

[unreleased]: https://github.com/perriera/extras_mon/compare/v0.7.0...HEAD
[0.7.0]: https://github.com/perriera/extras_mon/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/perriera/extras_mon/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/perriera/extras_mon/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/perriera/extras_mon/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/perriera/extras_mon/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/perriera/extras_mon/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/perriera/extras_mon/releases/tag/v0.1.0


test/monitor/mock_monitor.cpp
/**
 * @brief the "MIT/X Consortium License", (adapted for EXPARX.COM)
 *
 * Copyright (C) November 22, 2021 EXPARX INCORPORATED
 *
 * Permission is hereby  granted,  free of charge,  to  any  person
 * obtaining a copy of this software and  associated  documentation
 * files   (the  "Software"),  to deal  in   the  Software  without
 * restriction, including  without  limitation the rights  to  use,
 * copy,  modify, merge,  publish,  distribute,  sublicense, and/or
 * sell copies of the  Software, and to permit persons  to whom the
 * Software  is  furnished to  do  so,  subject  to  the  following
 * conditions:
 *
 * (See LICENSE.md for complete details)
 *
 */

#include <iostream>
#include <extras_mon/monitor.hpp>

#include "../vendor/catch.hpp"
#include "../vendor/fakeit.hpp"

using namespace extras;
using namespace fakeit;

SCENARIO("Mock MonitorInterface", "[MonitorInterface]") {

    bool wasTriggered = false;
    bool wasExecuted = false;
    Mock<mon::MonitorInterface> mock;
    When(Method(mock, event)).AlwaysDo([&wasExecuted]() {
        wasExecuted = true;
        });
    mon::MonitorInterface& i = mock.get();
    When(Method(mock, trigger)).AlwaysDo([&wasTriggered, &i]() {
        wasTriggered = true;
        i.event();
        });

    REQUIRE(wasTriggered == false);
    REQUIRE(wasExecuted == false);
    i.trigger();
    REQUIRE(wasTriggered == true);
    REQUIRE(wasExecuted == true);
    Verify(Method(mock, event));
    Verify(Method(mock, trigger));
}

test/monitor/test_monitor.cpp
/**
 * @brief the "MIT/X Consortium License", (adapted for EXPARX.COM)
 *
 * Copyright (C) November 22, 2021 EXPARX INCORPORATED
 *
 * Permission is hereby  granted,  free of charge,  to  any  person
 * obtaining a copy of this software and  associated  documentation
 * files   (the  "Software"),  to deal  in   the  Software  without
 * restriction, including  without  limitation the rights  to  use,
 * copy,  modify, merge,  publish,  distribute,  sublicense, and/or
 * sell copies of the  Software, and to permit persons  to whom the
 * Software  is  furnished to  do  so,  subject  to  the  following
 * conditions:
 *
 * (See LICENSE.md for complete details)
 *
 */

#include <iostream>
#include <extras_mon/monitor.hpp>

#include "../vendor/catch.hpp"
#include "../vendor/fakeit.hpp"

using namespace extras;

/**
 * @brief Monitor
 *
 */
concrete struct SampleMonitor implements mon::MonitorInterface
{

    bool wasTriggered = false;
    bool wasExecuted = false;

    virtual void event()  override {
        wasExecuted = true;
    }
    virtual void trigger()  override {
        wasTriggered = true;
        event();
    }

};

SCENARIO("Test MonitorInterface", "[MonitorInterface]") {

    SampleMonitor sample;
    mon::MonitorInterface& i = sample;

    REQUIRE(sample.wasTriggered == false);
    REQUIRE(sample.wasExecuted == false);
    i.trigger();
    REQUIRE(sample.wasTriggered == true);
    REQUIRE(sample.wasExecuted == true);
}

test/chessgame/mock_ChessGame.cpp
/**
 * @brief the "MIT/X Consortium License", (adapted for EXPARX.COM)
 *
 * Copyright (C) November 22, 2021 EXPARX INCORPORATED
 *
 * Permission is hereby  granted,  free of charge,  to  any  person
 * obtaining a copy of this software and  associated  documentation
 * files   (the  "Software"),  to deal  in   the  Software  without
 * restriction, including  without  limitation the rights  to  use,
 * copy,  modify, merge,  publish,  distribute,  sublicense, and/or
 * sell copies of the  Software, and to permit persons  to whom the
 * Software  is  furnished to  do  so,  subject  to  the  following
 * conditions:
 *
 * (See LICENSE.md for complete details)
 *
 */

#include <iostream>
#include <extras_mon/game/ChessGame.hpp>

#include "../vendor/catch.hpp"
#include "../vendor/fakeit.hpp"

using namespace extras;
using namespace fakeit;

SCENARIO("Mock ChessGameInterface: toOctal", "[CHES-9]") {

    Mock<mon::ChessGameInterface> mock;
    When(Method(mock, moves)).Return();

    mon::ChessGameInterface& i = mock.get();
    i.moves();
    Verify(Method(mock, moves));
}

mkpkg.py
"""
A simple script to create a debian package
using checkinstall.

Author: Matt Williams
"""

import argparse
import os
import subprocess as sp
import sys
from pathlib import Path

# Paranoid platform check
if sys.platform != "linux":
    sys.stderr.write(f"{sys.platform}' is currently not supported\n")
    sys.exit(1)

# check if user is root
if os.getenv("USER") == "root":
    sys.stderr.write(
        "\033[1;31mError\033[0m We humbly request that you do not run this script as root \n"
    )
    sys.exit(1)


class Colors:
    """
    Struct to represent colors
    """

    BBlack = "\033[1;30m"
    BRed = "\033[1;31m"
    BGreen = "\033[1;32m"
    BYellow = "\033[1;33m"
    BBlue = "\033[1;34m"
    BPurple = "\033[1;35m"
    BCyan = "\033[1;36m"
    BWhite = "\033[1;37m"
    Reset = "\033[0m"


class Packager:
    """
    A simple class that uses checkinstall to build
    a debian package
    """

    PKGNAME = "extras_mon"

    def __init__(self, args: argparse.Namespace):
        self.pkg_root_dir = Path(__file__).resolve(strict=True).parent
        self.build_dir = Path(f"{self.pkg_root_dir}") / "build"
        self.shell: str = os.getenv("SHELL")
        self.pkg_version: str = args.version
        self.pkg_license = "MIT"
        self.pkg_maintainer = "'mattcoding4days perriera'"

    @staticmethod
    def error(message: str) -> None:
        """
        outputs colorized error message and exits
        :return: None
        """
        sys.stderr.write(
            f"{Colors.BRed}ERROR{Colors.Reset}  {Colors.BWhite}{message}{Colors.Reset}\n"
        )
        sys.exit(1)

    @staticmethod
    def warn(message: str) -> None:
        """
        outputs colorized warning message
        :return: None
        """
        sys.stdout.write(
            f"{Colors.BYellow}WARNING{Colors.Reset}  {Colors.BWhite}{message}{Colors.Reset}\n"
        )

    @staticmethod
    def info(message: str) -> None:
        """
        outputs colorized info message
        :return: None
        """
        sys.stdout.write(
            f"{Colors.BGreen}INFO{Colors.Reset}  {Colors.BBlue}{message}{Colors.Reset}\n"
        )

    def __exec_command(self, command: str) -> None:
        """
        wrapper around subproccess, if a command fails to execute properly,
        error_msg will print, and promptly exit the entire script.
        :return: None
        """
        self.info(f"Running: {command}")
        try:
            sp.run(command, check=True, shell=True, executable=self.shell)
            self.info("Done.. \n")
        except sp.CalledProcessError as error:
            self.error(f"{error}")

    def __is_installed(self, command: str) -> bool:
        """
        wrapper around subproccess, to test if a program is already installed,
        this will not work for libraries etc, but only for executables
        :return: True, False
        """
        cmd = f"command -v {command}"
        ret: sp.CompletedProcess = sp.run(
            cmd, check=False, capture_output=True, shell=True
        )
        if ret.returncode != 0:
            # the program is not installed
            return False

        self.info(f"Found: {command}")
        return True

    def build_pkg(self) -> None:
        """
        build the package with parameters that were fed to us
        """
        if self.pkg_root_dir.stem != Packager.PKGNAME:
            self.error("Please run the script from project root")

        if not self.build_dir.exists():
            self.error("Please compile the library in a build directory first")

        self.info("Found: build dir")
        if not self.__is_installed("checkinstall"):
            self.error("checkinstall is not installed on the system")

        self.__exec_command(
            f"cd {str(self.build_dir)} && sudo checkinstall -y --pkgname={self.PKGNAME}"
            f" --pkgversion={self.pkg_version} --pkglicense={self.pkg_license}"
            f" --maintainer={self.pkg_maintainer} --install=yes"
        )


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        prog="mkpkg",
        description=f"Create a debian package for the {Packager.PKGNAME} library",
    )
    parser.add_argument(
        "version",
        type=str,
        help=f"The current version of the {Packager.PKGNAME} library",
    )

    arguments = parser.parse_args()
    package = Packager(arguments)
    package.build_pkg()


src/game/ChessGame.cpp
/**
 * @brief the "MIT/X Consortium License", (adapted for EXPARX.COM)
 *
 * Copyright (C) November 22, 2021 EXPARX INCORPORATED
 *
 * Permission is hereby  granted,  free of charge,  to  any  person
 * obtaining a copy of this software and  associated  documentation
 * files   (the  "Software"),  to deal  in   the  Software  without
 * restriction, including  without  limitation the rights  to  use,
 * copy,  modify, merge,  publish,  distribute,  sublicense, and/or
 * sell copies of the  Software, and to permit persons  to whom the
 * Software  is  furnished to  do  so,  subject  to  the  following
 * conditions:
 *
 * (See LICENSE.md for complete details)
 *
 */

#include <extras_mon/game/ChessGame.hpp>
#include <extras/interfaces.hpp>
#include <iostream>

using namespace std;
using namespace extras;

void mon::ChessGame::moves() const {}



include/extras_mon/monitor.hpp
/**
 * @file Monitor.hpp
 * @author Perry Anderson (perry@exparx.com)
 * @brief MonitorInterface
 * @version 0.1
 * @date 2021-11-30
 *
 * @copyright (C) November 22, 2021 EXPARX INCORPORATED
 *
 */

#ifndef _EXTRAS_MONITOR_HPP
#define _EXTRAS_MONITOR_HPP

 /**
  * @brief the "MIT/X Consortium License", (adapted for EXPARX.COM)
  *
  * Copyright (C) November 22, 2021 EXPARX INCORPORATED
  *
  * Permission is hereby  granted,  free of charge,  to  any  person
  * obtaining a copy of this software and  associated  documentation
  * files   (the  "Software"),  to deal  in   the  Software  without
  * restriction, including  without  limitation the rights  to  use,
  * copy,  modify, merge,  publish,  distribute,  sublicense, and/or
  * sell copies of the  Software, and to permit persons  to whom the
  * Software  is  furnished to  do  so,  subject  to  the  following
  * conditions:
  *
  * (See LICENSE.md for complete details)
  *
  */

#include <iostream>
#include <extras/interfaces.hpp>

namespace extras {
    namespace mon {

        /**
         * @brief MonitorInterface
         *
         */
        interface MonitorInterface
        {

            /**
             * @brief event
             * @note will be called from the trigger()
             */
            virtual void event() pure;

            /**
             * @brief trigger()
             * @note will determine if the event needs to be called
             *
             */
            virtual void trigger() pure;

        };

        /**
         * @brief Monitor
         *
         */
        concrete class Monitor implements MonitorInterface
        {

            virtual void event()  override;
            virtual void trigger()  override;

        };

        /**
         * @brief EndOfGameReachedException
         *
         */

        concrete class EndOfGameReachedException
            extends extras::AbstractCustomException {
        public:
            EndOfGameReachedException(
                const std::string& msg,
                const extras::WhereAmI& whereAmI)
                : AbstractCustomException(msg.c_str(),
                    whereAmI._file.c_str(),
                    whereAmI._func.c_str(), whereAmI._line) {}
            static void assertion(
                int sizePGN, int sizeFEN,
                const std::string& msg,
                const extras::WhereAmI& ref);
        };

    } // end namespace 
}

#endif // _EXTRAS_MONITOR_HPP


include/extras_mon/game/ChessGame.hpp
/**
 * @file ChessGame.hpp
 * @author Perry Anderson (perry@exparx.com)
 * @brief Sample Header file
 * @version 0.1
 * @date 2021-11-30
 *
 * @copyright (C) November 22, 2021 EXPARX INCORPORATED
 *
 */

#ifndef _CHESSGAME_HPP
#define _CHESSGAME_HPP

 /**
  * @brief the "MIT/X Consortium License", (adapted for EXPARX.COM)
  *
  * Copyright (C) November 22, 2021 EXPARX INCORPORATED
  *
  * Permission is hereby  granted,  free of charge,  to  any  person
  * obtaining a copy of this software and  associated  documentation
  * files   (the  "Software"),  to deal  in   the  Software  without
  * restriction, including  without  limitation the rights  to  use,
  * copy,  modify, merge,  publish,  distribute,  sublicense, and/or
  * sell copies of the  Software, and to permit persons  to whom the
  * Software  is  furnished to  do  so,  subject  to  the  following
  * conditions:
  *
  * (See LICENSE.md for complete details)
  *
  */

#include <iostream>
#include <extras/interfaces.hpp>

namespace extras {
    namespace mon {

        /**
         * @brief ChessGameInterface
         *
         */
        interface ChessGameInterface
        {

            /**
             * @brief moves()
             * @return all the chess moves of the given chess game
             */
            virtual void moves() const pure;

        };

        /**
         * @brief ChessGame
         *
         */
        concrete class ChessGame implements ChessGameInterface
        {

            /**
             * @brief moves()
             * @return all the chess moves of the given chess game
             */
            virtual void moves() const override;

        };

        /**
         * @brief EndOfGameReachedException
         *
         */

        concrete class EndOfGameReachedException
            extends extras::AbstractCustomException {
        public:
            EndOfGameReachedException(
                const std::string& msg,
                const extras::WhereAmI& whereAmI)
                : AbstractCustomException(msg.c_str(),
                    whereAmI._file.c_str(),
                    whereAmI._func.c_str(), whereAmI._line) {}
            static void assertion(
                int sizePGN, int sizeFEN,
                const std::string& msg,
                const extras::WhereAmI& ref);
        };

    } // end namespace 
}

#endif // _CHESSGAME_HPP


include/extras_mon/version.hpp
/**
* @file version.hpp
* @author Matt Williams (mattltf@protonmail.com)
* @brief Adds version support for Cmake script
* @version 3.2.0
* @date 2021-08-08
*
* @copyright (C) August 8, 2021 Matt Williams
*
*/

#ifndef _EXTRAS_MON_VERSION_HPP
#define _EXTRAS_MON_VERSION_HPP

/**
 * @brief the "MIT/X Consortium License", (adapted for EXPARX.COM)
 *
 * Copyright (C) November 22, 2021 EXPARX INCORPORATED
 *
 * Permission is hereby  granted,  free of charge,  to  any  person
 * obtaining a copy of this software and  associated  documentation
 * files   (the  "Software"),  to deal  in   the  Software  without
 * restriction, including  without  limitation the rights  to  use,
 * copy,  modify, merge,  publish,  distribute,  sublicense, and/or
 * sell copies of the  Software, and to permit persons  to whom the
 * Software  is  furnished to  do  so,  subject  to  the  following
 * conditions:
 *
 * (See LICENSE.md for complete details)
 *
 */

#define EXTRAS_MON_VER_MAJOR 0
#define EXTRAS_MON_VER_MINOR 7
#define EXTRAS_MON_VER_PATCH 0

#define EXTRAS_MON_VERSION \
  (EXTRAS_MON_VER_MAJOR * 10000 + EXTRAS_MON_VER_MINOR * 100 + EXTRAS_MON_VER_PATCH)

#endif// _EXTRAS_MON_VERSION_HPP
